request-deferrer/service-worker.js

```diff
@@ -0,0 +1,182 @@
+importScripts('./lib/ServiceWorkerWare.js');
+importScripts('./lib/localforage.js');
+
+// Determine the root for the routes. I.e, if the Service Worker URL is
+// `http://example.com/path/to/sw.js`, then the root is
+// `http://example.com/path/to/`
+var root = (function() {
+  var tokens = (self.location + '').split('/');
+  tokens[tokens.length - 1] = '';
+  return tokens.join('/');
+}());
+
+// By using Mozilla's ServiceWorkerWare we can quickly setup some routes
+// for a _virtual server_. **It is convenient you review the
+// [virtual server recipe](/virtual-server.html) before seeing this**.
+var worker = new ServiceWorkerWare();
+
+// So here is the idea. We will check if we are online or not. In case
+// we are not online, enqueue the request and provide a fake response.
+// Else, flush the queue and let the new request to reach the network.
+
+// This function factory does exactly that.
+function tryOrFallback(fakeResponse) {
+  // Return a handler that...
+  return function(req, res) {
+    // If offline, enqueue and answer with the fake response.
+    if (!navigator.onLine) {
+      console.log('No network availability, enqueuing');
+      return enqueue(req).then(function() {
+        // As the fake response will be reused but Response objects
+        // are one use only, we need to clone it each time we use it.
+        return fakeResponse.clone();
+      });
+    }
+
+    // If online, flush the queue and answer from network.
+    console.log('Network available! Flushing queue.');
+    return flushQueue().then(function() {
+      return fetch(req);
+    });
+  };
+}
+
+// A fake response with a joke for when there is no connection. A real
+// implementation could have cached the last collection of quotations
+// and keep a local model. For simplicity, not implemented here.
+worker.get(root + 'api/quotations?*', tryOrFallback(new Response(JSON.stringify([{
+  text: 'You are offline and I know it well.',
+  author: 'The Service Worker Cookbook',
+  id: 1,
+  isSticky: true
+}]), { headers: { 'Content-Type': 'application/json' } })));
+
+// For deletion, let's simulate that all went OK. **Notice we are omitting
+// the body of the response**. Trying to add a body with a 204, deleted, as
+// status throws an error.
+worker.delete(root + 'api/quotations/:id?*', tryOrFallback(new Response({
+  status: 204
+})));
+
+// Creation is another story. We can not reach the server so we can not
+// get the id for the new quotations. No problem, just say we accept the
+// creation and we will process it later, as soon as we recover connectivity.
+worker.post(root + 'api/quotations?*', tryOrFallback(new Response(null, {
+  status: 202
+})));
+
+// Lastly, this a virtual route to force the a sync. It is originated
+// by the client who is listening for the `online` event. This request
+// never reaches the network. It is here to be consistent but it is pure
+// virtual.
+worker.post(root + 'api/sync', function() {
```

Comment:
Yes, I mean, why do we need this synchronization? Couldn't the client directly send the GET request to load the quotations as soon as he is online?